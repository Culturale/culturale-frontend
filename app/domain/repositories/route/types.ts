export enum RouteState {
  FullySequenced = 'FULLY_SEQUENCED',
  NotFullySequenced = 'NOT_FULLY_SEQUENCED',
}
