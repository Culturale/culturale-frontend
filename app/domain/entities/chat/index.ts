export * from './chat';
export * from './chat.interface';
