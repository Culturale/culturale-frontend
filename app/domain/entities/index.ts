export * from './chat';
export * from './event';
export * from './message';
export * from './user';