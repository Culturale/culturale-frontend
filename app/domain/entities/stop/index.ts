export * from './stop';
export * from './stop.interface';
