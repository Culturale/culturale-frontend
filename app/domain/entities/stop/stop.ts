import { action, computed, makeObservable, observable } from 'mobx';

import type { IStop } from './stop.interface';

export class Stop implements IStop {
  public stopId: string;
  public address: string;
  public clientRadius?: number;
  public completed?: boolean;
  public retailerRadius?: number;
  public geoHash: string;

  constructor(stopId: string, address?: string, completed?: boolean) {
    this.stopId = stopId;
    this.completed = completed;
    this.address = address;

    makeObservable(this, {
      completeStop: action,
      completed: observable,
      isCompleted: computed,
      stopId: observable,
      uncompleteStop: action,
      updatePerimeter: action
    });
  }

  sequence?: number;

  public completeStop() {
    this.completed = true;
  }

  public uncompleteStop() {
    this.completed = false;
  }

  get isCompleted(): boolean {
    return this.completed;
  }

  public updatePerimeter(clientRadius: number, retailerRadius: number) {
    this.clientRadius = clientRadius;
    this.retailerRadius = retailerRadius;
  }

  public get sequenceNumber(): number {
    return 0;
  }
}
