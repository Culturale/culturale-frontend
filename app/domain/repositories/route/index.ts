export * from './route';
export * from './route.interface';
export * from './types';
