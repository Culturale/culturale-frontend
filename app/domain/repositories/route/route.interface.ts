import type { IStop } from '~/domain/entities';

import type { RouteState } from './types';

/**
 * Representation the route the driver has to follow
 *
 * @property route -> List of stops the driver has to do
 * @property routeState -> Indicates which is the state of the sequenced stops of the route
 */
export interface IRoute {
  route: IStop[];
  routeState: RouteState;

  /**
   * Updates the route with the provided list of stops
   * @public
   */
  setRoute: (route: IStop[]) => void;

  /**
   * Updates the route state with the provided value
   * @public
   */
  setRouteState: (state: RouteState) => void;

  /**
   * Returns the stop of the route with id
   * @public
   */
  getStop: (id: string) => IStop;

  /**
   * Returns true if the route is correctly sequenced
   * @readonly
   */
  readonly isRouteSequenced: boolean;

  /**
   * Returns all the stops that are completed
   * @readonly
   */
  readonly completedStops: IStop[];

  /**
   * Returns all the stops without being overlapped by geocode
   * @readonly
   */
  readonly visibleStops: IStop[];
}
