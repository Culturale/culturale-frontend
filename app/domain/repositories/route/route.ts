import { makeAutoObservable } from 'mobx';

import type { IStop } from '~/domain/entities/stop';

import type { IRoute } from './route.interface';
import { RouteState } from './types';

export class Route implements IRoute {
  public route: IStop[];
  public routeState: RouteState;

  constructor() {
    this.route = [];
    this.routeState = RouteState.FullySequenced;

    makeAutoObservable(this);
  }

  public setRoute(route: IStop[]) {
    this.route = route;
  }

  public setRouteState(state: RouteState) {
    this.routeState = state;
  }

  public getStop(id: string): IStop {
    return this.route.find((stop) => stop.stopId === id) || null;
  }

  get isRouteSequenced(): boolean {
    return this.routeState === RouteState.FullySequenced;
  }

  get visibleStops(): IStop[] {
    const stops = this.route;
    const visibleStops: IStop[] = [];

    for (const stop of stops) {
      const index = visibleStops.findIndex((s) => s.geoHash === stop.geoHash);

      if (index !== -1) {
        const stopWithSameGeoHash = visibleStops[index];
        const needsToBeReplaced =
          stopWithSameGeoHash.sequenceNumber > stop.sequenceNumber;

        if (needsToBeReplaced) {
          visibleStops[index] = stop;
        }
      } else {
        visibleStops.push(stop);
      }
    }

    return visibleStops;
  }

  get completedStops(): IStop[] {
    return this.route.filter((stop) => stop.isCompleted);
  }
}
