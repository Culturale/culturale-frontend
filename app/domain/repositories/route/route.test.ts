import { Stop } from '~/domain/entities/stop';

import { Route } from './route';
import type { IRoute } from './route.interface';
import { RouteState } from './types';

describe('Route', function () {
  let instance: IRoute;

  beforeEach(function () {
    instance = new Route();
  });

  it('can be created', function () {
    expect(instance).toMatchSnapshot();
  });

  describe('setRoute', function () {
    it('sets the route property correctly', function () {
      instance.setRoute([new Stop('test-id')]);

      expect(instance).toMatchSnapshot();
    });
  });

  describe('setRouteState', function () {
    it('sets the routeState property correctly', function () {
      instance.setRouteState(RouteState.NotFullySequenced);

      expect(instance.routeState).toBe(RouteState.NotFullySequenced);
    });
  });

  describe('getStop', function () {
    beforeEach(function () {
      instance.setRoute([new Stop('test-id')]);
    });

    it('returns null if no stop of the route matches the id provided', function () {
      const result = instance.getStop('fake-id');

      expect(result).toBeNull();
    });

    it('returns the stop matching the id provided if it exists', function () {
      const result = instance.getStop('test-id');

      expect(result).toBe(instance.route[0]);
    });
  });

  describe('get isRouteSequenced', function () {
    beforeEach(function () {
      instance.setRoute([new Stop('test-id')]);
    });

    it('returns true if route state is fully sequenced', function () {
      expect(instance.isRouteSequenced).toBeTruthy();
    });

    it('returns false if route state is not fully sequenced', function () {
      instance.setRouteState(RouteState.NotFullySequenced);

      expect(instance.isRouteSequenced).toBeFalsy();
    });
  });
});
