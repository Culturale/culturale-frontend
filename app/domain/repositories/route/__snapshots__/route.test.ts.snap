// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Route can be created 1`] = `
Route {
  "route": Array [],
  "routeState": "FULLY_SEQUENCED",
}
`;

exports[`Route setRoute sets the route property correctly 1`] = `
Route {
  "route": Array [
    Stop {
      "address": undefined,
      "clientRadius": undefined,
      "completed": undefined,
      "geoHash": undefined,
      "retailerRadius": undefined,
      "sequence": undefined,
      "stopId": "test-id",
    },
  ],
  "routeState": "FULLY_SEQUENCED",
}
`;
